Fix buffer overrun in ReadDERFromFile

sfc-gh-emusser opened the pull request:
This commit fixes an issue in ReadDERFromFile where it may overrun
the buffer containing the certificate bundle file when searching
for embedded certificates. In some cases, remnants of a previously
read certificate may be found past the buffer resulting in an error
improperly being returned.

The issue is that ReadDERFromFile solely relies on strstr returning
a nullptr when searching for the beginning of the next certificate.
Without a null terminator strstr is free search beyond the intended
buffer. The issue was readily seen by making a curl request in a loop.

The fix is add a null terminator at the end of the read buffer. A
check is also added that we are not beyond the end of the buffer.

sfc-gh-emusser committed:
Fix buffer overrun in ReadDERFromFile

This commit fixes an issue in ReadDERFromFile where it may overrun
the buffer containing the certificate bundle file when searching
for embedded certificates. In some cases, remnants of a previously
read certificate may be found past the buffer resulting in an error
improperly being returned.

The issue is that ReadDERFromFile solely relies on strstr returning
a nullptr when searching for the beginning of the next certificate.
Without a null terminator strstr is free search beyond the intended
buffer. The issue was readily seen by making a curl request in a loop.

The fix is add a null terminator at the end of the read buffer. A
check is also added that we are not beyond the end of the buffer.

kdudka commented on src/util.c:
```diff
@@ -82,14 +85,18 @@ static SECStatus FileToItem(SECItem * dst, PRFileDesc * src)
 	if (bytesReadNow == 0) {
 	    /* EOF */
-	    dst->len = bytesReadTotal;
+	    dst->len = bytesReadTotal + 1;
```
Either this should be `dst->len = bytesReadTotal` in case we do not want to include the terminating `NUL` in the binary data, or the next line should be `dst->data[bytesRedaTotal] = '\0'`, so that we write `NUL` right after the read data?

sfc-gh-emusser replied:
Good catch! I think setting `dst->len = bytesReadTotal; dst->data[bytesReadTotal] = '\0';` is probably the way to go.

One question I had is whether `dst->len` should instead contain the length of the buffer allocated as `SECITEM_ReallocItemV2` uses len to determine whether a reallocation is necessary. But in this case I see we just call `free` on the buffer when we are done so it doesn't matter. Should it ultimately be calling `SECITEM_FreeItem` ?

sfc-gh-emusser replied:
Looks like SECITEM_ReallocItemV2 doesn't maintain the actual allocation size.

kdudka replied:
Not that I found it documented anywhere but I believe that `SECItem::len` is the actual size of valid data to be processed.

kdudka reviewed (changes requested):
Thank you for the thorough analysis and implementing the fix!  One technical question inline though.

sfc-gh-emusser committed:
Fix null termination of buffer

This fixes a small issue in the prior buffer overrun fix where we adding
an extra byte before the null terminator.

kdudka commented:
Merging, thanks for the contribution!
